File: backend_aks_ide/src/controller/Authentication/checkEmail.ts
import { Request, Response } from "express";

import { prisma } from "../../prismaDb/prismaDb";

const checkEmail = async (req: Request, res: Response): Promise<void> => {
  try {
    const { email } = req.body;

    const user = await prisma.emailVerification.findUnique({
      where: {
        email: email,
      },
    });

    if (!user) {
      res.status(404).send("User not found");
      return;
    }

    if (user?.verified) {
      res
        .status(200)
        .send({ message: "Email already verified", Route: "PASSWORD" });
      return;
    }

    res.status(200).send({ message: "Email not verified", Route: "SENT" });
    return;
  } catch (error) {
    console.log("unable to check email", error);
    res.status(500).send("Unable to check email");
  }
};

export default checkEmail;

File: backend_ide/src/handleres/authenticate/check_email.rs
use axum::{
    extract::{Json, State},
    http::StatusCode,
};
use sea_orm::{ColumnTrait, EntityTrait, QueryFilter};
use serde::{Deserialize, Serialize};

use crate::{entities::email_verification, AppState};

#[derive(Debug, Deserialize)]
pub struct CheckEmailRequest {
    pub email: String,
}

#[derive(Debug, Serialize)]
pub struct EmailVerificationResponse {
    pub success: bool,
    pub message: String,
    pub route: Option<String>,
    pub error: Option<String>,
}

pub async fn check_email(
    State(state): State<AppState>,
    Json(payload): Json<CheckEmailRequest>,
) -> (StatusCode, Json<EmailVerificationResponse>) {
    match email_verification::Entity::find()
        .filter(email_verification::Column::Email.eq(&payload.email))
        .one(&*state.db)
        .await
    {
        Ok(Some(user)) => {
            if user.verified {
                (
                    StatusCode::OK,
                    Json(EmailVerificationResponse {
                        success: true,
                        message: "Email already verified".to_string(),
                        route: Some("PASSWORD".to_string()),
                        error: None,
                    }),
                )
            } else {
                (
                    StatusCode::OK,
                    Json(EmailVerificationResponse {
                        success: true,
                        message: "Email not verified".to_string(),
                        route: Some("SENT".to_string()),
                        error: None,
                    }),
                )
            }
        }
        Ok(None) => (
            StatusCode::NOT_FOUND,
            Json(EmailVerificationResponse {
                success: false,
                message: "User not found".to_string(),
                route: None,
                error: None,
            }),
        ),
        Err(e) => {
            eprintln!("Unable to check email: {}", e);
            (
                StatusCode::INTERNAL_SERVER_ERROR,
                Json(EmailVerificationResponse {
                    success: false,
                    message: "Unable to check email".to_string(),
                    route: None,
                    error: Some(e.to_string()),
                }),
            )
        }
    }
}